### PR TITLE
Structured maintainers

### PR DESCRIPTION
**Pull request by moretea**
###### Motivation for this change

Enables having proper metadata about maintainers, which is useful for https://github.com/NixOS/rfcs/pull/19 (maintainers file) / https://github.com/NixOS/rfcs/pull/21 (PR's on Hydra).

###### Example
```
nix-instantiate --strict --eval . -A pkgs.keen4.meta.maintainers
[ { GitHub = "edolstra"; IRC = "niksnut"; email = "eelco.dolstra@logicblox.com"; name = "Eelco Dolstra"; } ]
```

###### Things done
Verified that not everything is completely broken by running:
```
nix-build nixos/tests/boot.nix 
nix-build nixos/tests/firefox.nix
```
###### Why the manipulation of the data structure at the bottom of the file?
If  `addGitHub`  would not be applied to the attrset, then  keys of the attrset  cannot be observed when accessing them via `pkgs.my_pkg.meta.maintainers`.

###### Remaining questions
- [ ] Is setting IRC to null if not defined desired behavior?

### CODE REVIEW

File: lib/attrsets.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ lib }:
+{ lib, ... }:
```
Comment (edolstra) [inline lib/attrsets.nix]: Why is this needed? `...` considered harmful.

**Comment by edolstra** [comment]
This seems overly complicated to me. Why do we need that all that funky `_structuredMaintainers` and `...` business? I would just change `lib/maintainers.nix` to return a attrset of attrsets.

**Comment by moretea** [comment]
I would very much prefer to only have that!

My reasoning for implementing a flag is based on:
1. This structured data is very useful for the RFCs 19&21 to ping the correct maintainers.
2. I assumed that hydra depends on `pkgs.my_pkg.meta.maintainers` to be a list of email addresses.

Having a flag will allows us to get to use this data before support is added to Hydra.

**Comment by moretea** [comment]
We discussed this on IRC, and we'll just move to the new format, after which @edolstra will fix Hydra.

**Comment by FRidh** [comment]
I would like to take this further (see https://github.com/NixOS/nixpkgs/pull/34842#issuecomment-364756384). @moretea did you use a script to generate the attr sets?

**Comment by 7c6f434c** [comment]
I think the feature has been implemented somehow, so any PRs about maintainer list structure should be discussed relative to what we have now.
